project_setup.py

```diff
@@ -0,0 +1,181 @@
+
+
+def rtd_project_guess(
+    repo_url:str,
+) -> str:
+    """Guesses the RTD project name from the repo URL."""
+    return Path(repo_url).stem
+
+
+def codecov_project_guess(
+    repo_url:str,
+) -> str:
+    """Guesses the codecov project name from the repo URL."""
+    return "/".join(Path(repo_url).parts[-2:]).replace(".git", "")
+
+
+def run_setup(
+    repo_url:str,
+    package_name:str,
+    repo:git.Repo = git.Repo(),
+) -> None:
+    """Helper to setup a new project based on this template.
+
+    It does the following things:
+    1. Updates the pyproject.toml and README.md files
+    2. Renames the package_name directory to a python safe version of the
+       repo name
+    3. Updates the package_name and <package_name> in all files to a python safe
+       version of the repo name
+
+    Args:
+        repo_url (str): The URL of the repo
+        package_name (str): The name of the package
+        repo (git.Repo): The git repo to use
+
+    Returns:
+        None
+    """
+    # Step 1: Update the pyproject.toml and README.md ==========================
+    rtd_guess = rtd_project_guess(repo_url)
+    codecov_guess = codecov_project_guess(repo_url)
+
+    pyproject_toml_replacement_pairs = [
+        (TEMPLATE_REPO_URL, repo_url),
+    ]
+    if replace_file_contents(
+        pyproject_toml_replacement_pairs,
+        Path(repo.working_dir) / Path("pyproject.toml"),
+    ):
+        print(UPDATED_PYPROJECT_TOML)
+
+    readme_replacement_pairs = [
+        (TEMPLATE_REPO_URL, repo_url),
+        (TEMPLATE_RTD, rtd_guess),
+        (TEMPLATE_CODECOV, codecov_guess),
+    ]
+    if replace_file_contents(
+        readme_replacement_pairs,
+        Path(repo.working_dir) / Path("README.md"),
+    ):
+        print(UPDATED_README)
+
+    # Step 2: Rename the package_name directory ================================
+    repo.git.mv(TEMPLATE_PACKAGE_NAME, package_name)
+    repo.git.commit("-m", f"renamed dir {TEMPLATE_PACKAGE_NAME} to {package_name}")
+    print(UPDATED_DIR.format(old_dir=TEMPLATE_PACKAGE_NAME, new_dir=package_name))
+
+    # Step 3: Update the package_name and <package_name> in all files ==========
+
+    # filter out files that we don't want to modify
+    files_to_check = list(filter(
+        valid_file_predicate,
+        map(lambda f: Path(repo.working_dir) / Path(f), repo.git.ls_tree("-r", "--name-only", "HEAD").split("\n")),
+    ))
+
+    # order matters, replace the bracketed version first
+    package_name_replacement_pairs = [
+        ("<" + TEMPLATE_PACKAGE_NAME + ">", package_name),
+        (TEMPLATE_PACKAGE_NAME, package_name),
+    ]
+    for file_path in files_to_check:
+        # only print positive results
+        if replace_file_contents(package_name_replacement_pairs, file_path):
+            print(UPDATED_FILE.format(package_name=package_name, file_path=file_path))
+
+    self_destruct(Path(repo.working_dir))
+
+    repo.git.commit("-am", f"updated package name to {package_name} in files")
```

Comment:
I am not committed to the second commit, but I need the first one for the git ls-tree to work.

I can drop if that is the preference of the group.

Reply:
could you do it in the other order? e.g. ls-tree to find all the files who need replacements, then git mv?

Reply:
I can, but I would need to edit the returned paths from `ls-tree` after the fact to reflect the `mv`. I'll try it out and see if it's as easy as it appears. It would be nice to avoid doing any commits.

Reply:
I don't think you'd have to edit any path. Just do all replacements and then last thing do the mv, it should sort it out unless I'm mistaken

Reply:
Ah I see. I think you're right, I was just caught up in the order I thought about them in.

Reply:
@amitschang What's wrong with using ``-a`` here?

>-a, --all
           Tell the command to automatically stage files that have been modified and deleted, but new files you have not told Git about are not affected.

Reply:
I'm not sure that the order does matter, but my pref is on the two commits but only such that the ``git mv`` commit is clean and only shows renames and no actual deltas.